courseworks/week8/TianhaoHU/hw8.ipynb

```diff
@@ -0,0 +1,185 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": 68,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "using Optim"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 69,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "create_loss_function (generic function with 1 method)"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "function create_loss_function(vertex_count, edges, locations)\n",
+    "    function loss(x)\n",
+    "        total_loss = 0.0\n",
+    "        for i in 1:vertex_count\n",
+    "            for j in (i+1):vertex_count\n",
+    "                xi, xj = x[2*i-1:2*i], x[2*j-1:2*j]\n",
+    "                dist = sqrt(sum((xi - xj).^2))\n",
+    "                connected = any([(i, j) == edge || (j, i) == edge for edge in edges])\n",
+    "\n",
+    "                if connected && dist > 1.0\n",
+    "                    total_loss += (dist - 1.0)^2\n",
+    "                elseif !connected && dist < 1.0\n",
+    "                    total_loss += (1.0 - dist)^2\n",
+    "                end\n",
+    "            end\n",
+    "        end\n",
+    "        return total_loss\n",
+    "    end\n",
+    "end"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 70,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "unit_disk_embedding (generic function with 1 method)"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "function unit_disk_embedding(vertex_count, edges)\n",
+    "    locations = rand(2, vertex_count)\n",
+    "    loss_function = create_loss_function(vertex_count, edges, locations)\n",
+    "    res = optimize(loss_function, vec(locations), LBFGS())\n",
+    "    min_loss = Optim.minimum(res)\n",
+    "\n",
+    "    if min_loss < 1e-6\n",
+    "        locations_opt = reshape(Optim.minimizer(res), 2, vertex_count)\n",
+    "        return true, locations_opt\n",
+    "    else\n",
+    "        return false, nothing\n",
+    "    end\n",
+    "end"
```

Comment:
Here you only checked if the loss function has beed optimized, but you also need to check by the defination of the unitdisk embeding, becase they are not equivalent.